src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -464,37 +457,87 @@ private static void UpdateBindingsOfAssemblyBeingBuilt(ArrayBuilder<AssemblyRefe
             referenceBindings[0] = bindingsOfAssemblyBeingBuilt.ToArrayAndFree();
         }
 
-        internal AssemblyData ResolveMissingAssembly(
+        /// <summary>
+        /// Resolve <paramref name="referenceIdentity"/> using a given <paramref name="resolver"/>.
+        /// 
+        /// We make sure not to query the resolver for the same identity multiple times (across submissions).
+        /// Doing so ensures that we don't create multiple assembly symbols within the same chain of script compilations 
+        /// for the same implicitly resolved identity. Failure to do so results in cast errors like "can't convert T to T".
+        /// 
+        /// The method only records successful resolution results by updating <paramref name="implicitReferenceResolutions"/>.
+        /// Failures are only recorded after all resolution attempts have been completed.
+        /// 
+        /// This approach addresses teh following scenario. Consider a script:
+        /// <code>
+        ///   #r "dir1\a.dll"
+        ///   #r "dir2\b.dll"
+        /// </code>
+        /// where both a.dll and b.dll reference x.dll, which is present only in dir2. Let's assume the resolver first 
+        /// attempts to resolve "x" referenced from "dir1\a.dll". The resolver may fail to find the dependency if it only
+        /// looks up the directory containing the referencing assembly (dir1). If we recorded and this failure immediately
+        /// we would not call the resolver to resolve "x" within the context of "dir2\b.dll" (or any other referencing assembly). 
+        /// 
+        /// This behavior would ensure consistency and if the types from x.dll do leak thru to the script compilation, but it 
+        /// would result in a missing assembly error. By recording the failure after all resolution attempts are complete
+        /// we also achieve a consistent behavior but are able to bind the reference to "x.dll". Besides, this approach
+        /// also avoids dependency on the order in which we evaluate the assembly references.
```

Comment:
>also avoids dependency on the order in which we evaluate the assembly references. [](start = 12, length = 81)

Is this true? Can different folders have different assemblies of the same identity? Then, the order of folders to look in is significant.

Reply:
Right, in general, if there are multiple different versions of the assembly with the same identity then it indeed depends on the order. That usually doesn't happen for shipping libraries, but it can definitely happen for ad-hoc built libraries. The statement is made in the context of x.dll being only in one of the directories. Let me clarify that in the comment.

Reply:
Added clarification